src/hp/derp/clients.rs

```diff
@@ -0,0 +1,345 @@
+//! based on tailscale/derp/derp_server.go
+//!
+//! The "Server" side of the client. Uses the `ClientConnManager`.
+use crate::hp::key::node::PublicKey;
+use std::collections::HashMap;
+
+use tokio::sync::mpsc;
+
+use super::{
+    client_conn::{ClientConnManager, Packet, PeerConnState},
+    conn::Conn,
+};
+
+/// Number of times we try to send to a client connection before dropping the data;
+const RETRIES: usize = 3;
+
+/// Represents a connection to a client.
+///
+/// From goimpl:
+///
+/// "Represents one or more connections to a client
+///
+/// In the common cast, the client should only have one connection to the DERP server for a given
+/// key. When they're connected multiple times, we record their set of connection, and keep their
+/// connections open to make them happy (to keep them from spinning, etc) and keep track of whihc
+/// is the lastest connection. If only the last is sending traffic, that last one is the active
+/// connection and it gets traffic. Otherwise, in the case of a cloned node key, the whole set of
+/// connections doesn't receive data frames."
+// TODO: expand to allow for multiple connections associated with a single PublicKey. This
+// introduces some questions around which connection should be prioritized when forwarding packets
+#[derive(Debug)]
+struct Client<C>
+where
+    C: Conn,
+{
+    /// The set of all connections associated with the PublicKey
+    // conns: HashMap<usize, ClientConnManager<C>>,
+    /// The most recent edition to the set, or the last connection we have received data from. Can
+    /// be `None` if that particular connection has disconnected & we have no other previous
+    /// connections
+    // last: Option<usize>,
+    /// the "home" connection for the client.
+    /// TODO: expect for setting and unsetting, I do not see this used in the go impl.
+    preferred: Option<usize>,
+    /// the connection
+    conn: ClientConnManager<C>,
+}
+
+impl<C> Client<C>
+where
+    C: Conn,
+{
+    pub fn new(conn: ClientConnManager<C>) -> Self {
+        Self {
+            preferred: None,
+            conn,
+        }
+    }
+
+    pub async fn shutdown(self) {
+        self.conn.shutdown().await;
```

Comment:
Ok, so I originally had these shutdowns happen in a spawned task & left to run on its own, without blocking.

I thought it might be better to have those shutdown without blocking, since these will be called in the main server loop, meaning we could be waiting for the connection to shutdown instead of picking up the next task of work in the loop.

However, if we do that, we won't ever have a way to know if we've actually cleaned up all of our spawned tasks... so I made them async again, and we wait until the shutdown finishes before moving on.

Not sure if that's the correct trade-off, though, so would like some feedback.